hummingbot/connector/exchange/kucoin/kucoin_api_user_stream_data_source.py

```diff
@@ -60,50 +56,50 @@ async def get_listen_key(self):
                     data: Dict[str, str] = await response.json()
                     return data
 
-    async def _subscribe_topic(self, ws: websockets.WebSocketClientProtocol) -> AsyncIterable[str]:
+    async def _subscribe_topic(self, ws: aiohttp.ClientWebSocketResponse) -> AsyncIterable[str]:
         try:
-            for topic in KUCOIN_PRIVATE_TOPICS:
+            for topic in CONSTANTS.PRIVATE_ENDPOINT_NAMES:
                 subscribe_request = {
                     "id": int(time.time()),
                     "type": "subscribe",
                     "topic": topic,
                     "privateChannel": True,
                     "response": True}
                 async with self._throttler.execute_task(CONSTANTS.WS_REQUEST_LIMIT_ID):
-                    await ws.send(ujson.dumps(subscribe_request))
+                    await ws.send_json(subscribe_request)
         except asyncio.CancelledError:
             raise
         except Exception:
             return
 
-    async def get_ws_connection(self) -> websockets.WebSocketClientProtocol:
+    async def get_ws_connection(self, client: aiohttp.ClientSession) -> aiohttp.ClientWebSocketResponse:
         stream_url: str = f"{self._current_endpoint}?token={self._current_listen_key}&acceptUserMessage=true"
         self.logger().info(f"Connecting to {stream_url}.")
 
         async with self._throttler.execute_task(CONSTANTS.WS_CONNECTION_LIMIT_ID):
-            return await websockets.connect(stream_url, ping_interval=40, ping_timeout=self.PING_TIMEOUT)
+            return await client.ws_connect(stream_url, autoping=True, heartbeat=CONSTANTS.WS_PING_HEARTBEAT)
```

Comment:
With the `autoping` set to `True`, we would no longer update the `_last_recv_time` upon receiving a ping message. I think it would be best to set it to `False` since this would affect the polling interval in `Exchange/Derivative` class.

Reply:
Good point. I've added a ping-pong loop and also added a check of `last_recv_time` in the `status_dict` of the exchange as that was the only use for the property as far as I could tell.